ユーザー管理機能

evetakumatsu opened the pull request:
# what
ユーザー管理機能の実装
# why
+ `devise`を導入してユーザー管理機能を実装した。
+ ユーザー登録機能が正常に働いてるか確認するため、
`user`モデルの単体テストを行った。

___
### 必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
___
![alt text](https://gyazo.com/85cf4026814a7441c330a3dfb8af1514.gif)

___
### 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、  
### エラーメッセージが出力される動画
___
![alt text](https://gyazo.com/e5d88375df23d6ebe3daf57893ff81fd.gif)

___
### 必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
___
![alt text](https://gyazo.com/6f7833d1b5fb2da4db2a7f993dd8f0b8.gif)

___
### 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、  
### エラーメッセージが出力される動画
___
![alt text](https://gyazo.com/4a32981a64f5e4bdc55b1f55aa7303cb.gif)

___
### ログアウトに成功する動画
___
![alt text](https://gyazo.com/220650c3ca14845475bdc6c77dcffdc5.gif)
___
### `user`モデルテストに成功する画像
___
![alt text](https://gyazo.com/94afcc989f9020de1dc7160add1db3ff.png)

evetakumatsu committed:
deviseの導入

evetakumatsu committed:
userテーブルの作成

evetakumatsu committed:
usersテーブルのマイグレーション設定

evetakumatsu committed:
userモデルのバリデーションの記述

evetakumatsu committed:
deviseのビューファイルを作成

evetakumatsu committed:
ダウンロードしたdeviseのビューファイルの配置

evetakumatsu committed:
新規登録画面のカラムを変更

evetakumatsu committed:
新規登録画面のモデルとURLの指定

evetakumatsu committed:
ダウンロードしたapplication.html.erbファイルを配置

evetakumatsu committed:
トップページの背景画像を表示させる

トップページで背景画像が表示されていなかったので、修正する。（本来ならアセットパイプラインを有効にしたかったが、コードを追加しても反映されないし、サンプルのサイトはurlで実装されている為、この修正にした。）

evetakumatsu committed:
新規登録、ログイン画面、ログアウトのパスを設定

evetakumatsu committed:
ログインの有無で表示を変える設定

evetakumatsu committed:
不要なコメントアウトの削除

evetakumatsu committed:
ログイン画面のデータのカラムを設定

evetakumatsu committed:
パスワードのカラム名を修正

evetakumatsu committed:
ユーザー登録失敗時のエラーメッセージ

エラーメッセージを表示させる。

evetakumatsu committed:
デバッグツールの導入

binding.pryを使用できるようにする。

evetakumatsu committed:
userモデルのバリデーション設定修正

今の状態だと、パスワードが保存できないエラーが出るため、修正した。

evetakumatsu committed:
deviseのコントローラー設定

ユーザー新規登録時の受け取るpermitを設定。

evetakumatsu committed:
ログアウト時のHTTPメソッド修正

現状だとログアウトできない為、HTTPメソッドの指定方法を変更した。

evetakumatsu committed:
ログイン時のform_with設定

form_withの送信先のモデルとパスを設定した。

evetakumatsu committed:
モデルの単体テストコードに必要なGemを導入

Rspec、FactoryBot、FakerのGemを記述。それと行を修正した。

evetakumatsu committed:
Rspecをインストール

evetakumatsu committed:
モデルの単体テスト結果を分かりやすく可視化する設定

evetakumatsu committed:
userモデル単体テストファイル作成

単体テストに使うFactoryBotファイルとuser_specファイルを作成

evetakumatsu committed:
Fakerの生成を日本語に変更

evetakumatsu committed:
カタカナだけのダミーデータ作成の為のGem導入

gem 'gimei'を導入

evetakumatsu committed:
gem 'gimei'を有効にする

evetakumatsu committed:
userモデル単体テストに使うダミーデータの作成

ユーザー新規登録に使う仮のユーザー情報をFactoryBotで設定

evetakumatsu committed:
userモデルの単体テストコード作成

evetakumatsu committed:
不要なコメントアウトを削除

evetakumatsu committed:
Rubocopのスタイルチェック修正

指摘を受けた箇所を修正した。

sh10n7 commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+  VALID_KANA_NAME_REGEX = /\A[ァ-ヶー]+\z/
+
+  validates :nickname,           presence: true
+  validates :password,           presence: true, format: { with: VALID_PASSWORD_REGEX  }
```
```
validates :password, presence: true, format: { with: VALID_PASSWORD_REGEX }
```
``` presence: true```のバリデーションは削除しておきましょう！

理由：deviseがデフォルトで付与しているため。

evetakumatsu replied:
コードを修正しました。

sh10n7 commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -0,0 +1,34 @@
+<%= render "shared/second-header"%>
+
+<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
```
```
<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
```
urlを修正しましょう！
現状記述いただいているpathはログインページに遷移するurlかと存じます。
ログインを実施するためのpathを再度rails routesで確認し、記載しましょう。

evetakumatsu replied:
コードを修正しました。

sh10n7 commented on db/migrate/20230728074004_devise_create_users.rb:
```diff
@@ -0,0 +1,50 @@
+      t.string   :first_name,         null: false
+      t.string   :kana_last_name,     null: false
+      t.string   :kana_first_name,    null: false
+      t.datetime :date_of_birth,      null: false
```
```
t.datetime :date_of_birth, null: false
```
生年月日のカラムでは、年月日の保存を行うため、カラム型にはdate型を指定しましょう。

evetakumatsu replied:
コードを修正しました。

sh10n7 commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname              { Faker::Name.last_name }
+    email                 { Faker::Internet.email }
+    password              { Faker::Internet.password(min_length: 6, mix_case: false) }
```
```
password { Faker::Internet.password(min_length: 6, mix_case: false) }
```
◎passwordカラムを修正しましょう。
現状、必ずしも半角英数混合になるかどうかが不明となる記述です。
(Fakerのみの場合、英字のみのパスワードが生成されてしまうことがあるためです。)
確実に半角英数混合の値を生成するため、『+』の記号を使用し、文字列を組み合わせることで期待する値を生成する内容に修正していきましょう。

evetakumatsu replied:
コードを修正しました。

sh10n7 reviewed (commented):
レビュー依頼ありがとうございます。
修正いただきたい箇所をコメントしております。ご確認いただけますと幸いです。
​
修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

その際、**修正後のテスト実行結果の画像**も添付いただきますようお願いいたします。

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞ、よろしくお願いいたします。

evetakumatsu committed:
userモデルのバリデーション設定修正

presence: trueの削除

evetakumatsu committed:
ログインページのformのURL設定修正

URLを送信するPOSTメソッドに修正した。

evetakumatsu committed:
マイグレーションファイルにて、データの型修正

生年月日のカラムをdate型に修正

evetakumatsu committed:
ダミーデータの設定修正

passwordを生成する時、必ず数字が入るように修正。
最初は「＋」の記号を使用したが、Rubocopのスタイルチェックで指摘を受けたため、文字列補間設定を使用した。

evetakumatsu committed:
Rubocopのスタイルチェック修正

指摘を受けた部分を修正した。

mihot117 commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname              { Faker::Name.last_name }
+    email                 { Faker::Internet.email }
+    password              { "#{Faker::Internet.password(min_length: 6, mix_case: false)}1" }
```
```ruby
{ "#{Faker::Internet.password(min_length: 6, mix_case: false)}1" }
```

上記ですと`#{Faker::Internet.password(min_length: 6, mix_case: false)}`で出力される文字列と1が出力される記述になっていますが、
`#{Faker::Internet.password(min_length: 6, mix_case: false)}`は数値のみ出力されてしまう可能性がありますので、英数字混合の文字列が毎回表示される確率が100%ではありません。

```ruby
{ Faker::Internet.password(min_length: 6, mix_case: false) + "1a"}
```

上記のように設定いただくことで、Fakerでどのような値が生成されても1aが付くので安心です。

evetakumatsu replied:
コードを修正しました。

mihot117 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。

追加でコメントいたしましたので、ご確認いただけますと幸いです。
factorybotの修正部分に対してコメントを再度残しています。



修正後、再度レビュー依頼フォームをご提出ください。

ご対応のほど、よろしくお願いいたします。

evetakumatsu committed:
ダミーデータの設定再修正

パスワードに`1a`が必ず含まれるように修正した。

sh10n7 reviewed (commented):
修正とレビュー依頼ありがとうございます。

修正点確認いたしました。
しっかりと実装できていることが確認できましたので、**LGTM**とさせていただきます！
実装お疲れ様でした。

こちらのブランチは忘れずにMergeを行っていただき、ローカルのmasterブランチにPullしましょう。

その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
ここまでの実装、本当にお疲れ様でした。
引き続き、応援しております！
